### DISCUSSION TITLE
Switching displays option

### DISCUSSION THREAD

**Discussion by ovidiubarbu** [appium/appium#19639] [Q&A]
Hi,

I have an Android device with output on 2 displays. Right now, I can see the `display 0` ( which seems to be the default), but could not find a way to see `display 1` in Appium Inspector. Is there a way to do that?

Any help would be appreciated!

**Comment by jlipps** [comment]
I don't think Appium has the ability to switch which display is used for automation/screenshot. Or do we @mykola-mokhnach?

**Comment by ovidiubarbu** [comment]
Would appreciate if this will be considered for implementation, if not there already in some way. Nowadays, a lot of car infotainment systems have more than 1 screen, and would help our testing. Also, if you could point to the code that could adjusted to support this would be awesome.

**Reply by KazuCocoa** [reply]
So far, https://github.com/appium/appium-uiautomator2-driver?tab=readme-ov-file#mobile-screenshots can specify display id.

I saw that the latest uiautomator library 2.3.0-beta01 has display id implementation in the backend, which did not have in the past. e.g., https://github.com/appium/appium-uiautomator2-server/pull/591; thus, potentially, uiautomator library will provide multiple display's test availability in the future.

Not sure how the library currently provides such functionality, thus it is appreciate to share your experience to achieve multi-display automation with uiautomator in Android. Then, perhaps appium uiautomator2 driver will have such functionality in the future. (then, maybe https://github.com/appium/appium-uiautomator2-server needs to be refactored/rewrite)

**Comment by ovidiubarbu** [comment]
Could not really find a solution yet. What I managed to do is to take a screenshot of the display 1, and an XML dump using adb commands. I am wondering if there is a way to feed them to the inspector and or Appium manually, and maybe be able to work with them. Long shot, I guess.

**Reply by mykola-mokhnach** [reply]
do you know how the application under test works? Does it have multiple windows shown on different displays? Or there is only one window and you can set on which display it should be opened?

**Reply by mykola-mokhnach** [reply]
from what I saw in the recent UIAutomator source it tries to fetch the actual display number the main activity opens on (Display.DEFAULT_DISPLAY) and works with that as default (also takes screenshots from there). Although, if you need to interact with some element which is located on another display (and most likely in another application window) then the interaction should properly be done there if this element is chosen for interaction.

**Reply by mykola-mokhnach** [reply]
Linking https://androidx.tech/artifacts/test.uiautomator/uiautomator/2.3.0-beta01-source/androidx/test/uiautomator/UiDevice.java.html

**Reply by mykola-mokhnach** [reply]
@eglitise Perhaps we may add a new feature to the inspector, which would allow to show more than one display, but this does not look like a trivial task and also differs for different platforms

**Reply by eglitise** [reply]
I think the simplest way would be to have UiAutomator2 expose a method to switch the current display, which can then be mapped to a new button in the Inspector GUI. Having multiple displays shown at once could be a follow-up task.

**Reply by eglitise** [reply]
In case `mobile: screenshots` or the listed `adb` command actually does support this already, then yes, I guess this can be made into an Inspector feature request.

**Comment by ovidiubarbu** [comment]
This is how I did it via adb:

```
adb shell am start com.google.android.calculator --display 1
adb shell screencap -d 1 -p /sdcard/screencapture.png
adb pull /sdcard/screencapture.png ~/_test

adb shell uiautomator dump 
adb pull /sdcard/window_dump.xml ~/_test/ddisplay_1.xml
```

**Comment by ovidiubarbu** [comment]
I have requested a feature to support this here:
[https://github.com/appium/appium/issues/19662](https://github.com/appium/appium/issues/19662)
